### PR TITLE
Introduce generic Executor base class and add tests

### PR DESCRIPTION
**Pull request by beckerhe**
This is moving around a lot of stuff regarding executors.

1. There is now a new base class `orbit_base::Executor` which implements `Schedule`, `ScheduleAfter` and `ScheduleAfterIfSuccess`.
2. `MainThreadExecutor` and `ThreadPool` inherit from `Executor`.
3. There are now tests for Executor
4. There is now a `SimpleExecutor` which also implements `Executor` and can be used in tests. It's different to `ImmediateExecutor` in the sense that scheduled actions are stored until the user executes them manually by calling `ExecuteScheduledTasks`.

### CODE REVIEW

File: src/OrbitBase/SimpleExecutor.cpp
Comment (antonrohr) [inline src/OrbitBase/SimpleExecutor.cpp:31]: Why not `std::make_shared`?

File: src/OrbitBase/SimpleExecutorTest.cpp
Comment (antonrohr) [inline src/OrbitBase/SimpleExecutorTest.cpp:58]: Is `chained_future` already finished between these 2 lines?

File: src/OrbitBase/SimpleExecutorTest.cpp
Comment (antonrohr) [inline src/OrbitBase/SimpleExecutorTest.cpp:75]: At this point the lambda is still in `Executor::waiting_continuations`, correct? I wonder if checking the value of `GetNumberOfWaitingContinuations` here, is a good or a bad idea. I think it might help explaining whats going on to the reader of this test, but it also unnecessarily ties this test suite to the implementation of the `Executor` base class.

**Review by antonrohr — APPROVED** [review]
1. 
 > It might also be interesting for folks who want to use promises and
futures in parts of Orbit which can't depend on Qt (like the service).

Is that actually the case, or is `Immediate Executor` the better choice? And while we are at it, is it planned that `ImmediateExecutor` also inherits from `Executor`? 

2. 
Thank you for all the tests (it took me a minute to wrap my head around `SaveUniquePtr`). I generally like them a lot and how they only EXPECT things in the realm of the test. Opposed to checking every pre and post condition. I made some remarks to the tests that might very well be bad ideas, so please take them with a grain of salt. 

3. 
The code looks good to me, but I am not comfortable judging on a conceptual level, whether things could be done differently / better. Maybe @dimitry- could take a look

File: src/OrbitBase/SimpleExecutor.cpp
Comment (beckerhe) [reply src/OrbitBase/SimpleExecutor.cpp:31]: SimpleExecutor has a private constructor and make_shared/unique can't be used.

File: src/OrbitBase/SimpleExecutorTest.cpp
Comment (beckerhe) [reply src/OrbitBase/SimpleExecutorTest.cpp:58]: No. `promise.SetResult` schedules the continuation, and `ExecuteScheduledTasks` executes the continuation and completes the future. Does that make sense?

File: src/OrbitBase/SimpleExecutorTest.cpp
Comment (beckerhe) [reply src/OrbitBase/SimpleExecutorTest.cpp:75]: There is also another point to consider: Ideally we wanna have orthogonal tests, meaning whenever I introduce a single bug, ideally I want to have a single test fail. This will allow me to pinpoint the problem very quickly. Of course that's not always possible, or let's say most of the time that's very hard because of dependencies.

But whenever possible I prefer not to test behaviour that is already covered by another test. In this case it's covered by the `Executor.*` tests.

Needless to say that these tests are far away from consistently following this guideline because most of them are just copied from what has already been there. But I tried at least not to introduce more double-coverage.

Nevertheless, if you believe this test (these tests) requires some explanation for the reader I can totally improve on that. I guess in that case comments would be the way to go. WDYT?

File: src/OrbitBase/include/OrbitBase/Executor.h
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Executor.h]: good catch. done. thanks

File: src/OrbitBase/ExecutorTest.cpp
Comment (beckerhe) [reply src/OrbitBase/ExecutorTest.cpp:161]: good point. done.

File: src/OrbitBase/ExecutorTest.cpp
Comment (beckerhe) [reply src/OrbitBase/ExecutorTest.cpp:41]: yes. done.

File: src/OrbitBase/include/OrbitBase/Executor.h
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/Executor.h]: done. I hope I caught all occurrences.

File: src/OrbitBase/SimpleExecutorTest.cpp
Comment (antonrohr) [reply src/OrbitBase/SimpleExecutorTest.cpp:58]: I looked at the code for a while and then decided to just test it. 
The following "is true" (which I tried to ask earlier)
```
promise.SetResult(ErrorMessage{kErrorMessage});
EXPECT_TRUE(chained_future.IsFinished());
executor->ExecuteScheduledTasks();
```

I looked more at the code and it makes sense. This is the "Short Circuit Pathway" in `Executor::ScheduleAfterIfSuccess`. On `promise.SetResult` the continuations of the corresponding future are called and with `ScheduleAfterIfSuccess` this continuation "short circuits" and effectively also calls "`chained_promise.SetResult`" (promise of `chained_future`). In this case another action is scheduled to erase the old functor (`[&called]() { called = true; }`), which was never executed. 

This is all complicated stuff. But I think it all checks out in the end. Also maybe you misunderstood my initial question. 

Just for completion, this is how I augmented the test to understand it:
```
...
  auto future = promise.GetFuture();
  EXPECT_EQ(executor->GetNumberOfWaitingContinuations(), 0);
  auto chained_future = executor->ScheduleAfterIfSuccess(future, [&called]() { called = true; });
  EXPECT_FALSE(called);
  EXPECT_FALSE(chained_future.IsFinished());

  EXPECT_EQ(executor->GetNumberOfWaitingContinuations(), 1);
  constexpr const char* const kErrorMessage{"Error"};
  promise.SetResult(ErrorMessage{kErrorMessage});
  EXPECT_TRUE(chained_future.IsFinished());
  EXPECT_EQ(executor->GetNumberOfWaitingContinuations(), 1);
  executor->ExecuteScheduledTasks();
  EXPECT_EQ(executor->GetNumberOfWaitingContinuations(), 0);
  EXPECT_FALSE(called);
...
```

File: src/OrbitBase/SimpleExecutorTest.cpp
Comment (antonrohr) [reply src/OrbitBase/SimpleExecutorTest.cpp:75]: Using the term orthogonal, plus the explanation is a great way to put in words what was a somewhat fuzzy feeling in my head. I have to remember that. 

So yes, I agree, lets leave the tests as is. I just tried to come up with explaining comments, but at this point I think I fully understood all the code and then the tests also are clear the way they currently are. So im going to resolve the conversation

File: src/OrbitBase/SimpleExecutorTest.cpp
Comment (beckerhe) [reply src/OrbitBase/SimpleExecutorTest.cpp:58]: Ahh, yes. You're right. Sorry. I didn't get that this was about the error case.

**Review by antonrohr — APPROVED** [review]
I know I requested Dimitry's review earlier, but after letting it sit for a while and looking at it again, I now feel more comfortable that I fully understood it. Also its mainly moving things around. So feel free to merge even without another review

File: src/OrbitBase/SimpleExecutorTest.cpp
Comment (beckerhe) [reply src/OrbitBase/SimpleExecutorTest.cpp:58]: I quickly thought about it and I'm not entirely sure if that behaviour is an implementation detail or not. Nevertheless I will add the check! On first thought it makes sense to me, we can still remove it in the future!

Comment (beckerhe) [reply src/OrbitBase/SimpleExecutorTest.cpp]: Done.

**Comment by beckerhe** [comment]
> I know I requested Dimitry's review earlier, but after letting it sit for a while and looking at it again, I now feel more comfortable that I fully understood it. Also its mainly moving things around. So feel free to merge even without another review

Thanks. I just rebased and I will merge when the CI is done.

File: src/OrbitBase/ThreadPool.cpp
Comment (dimitry-) [inline src/OrbitBase/ThreadPool.cpp:197]: why did it become shared?

File: src/OrbitBase/SimpleExecutor.cpp
Comment (dimitry-) [inline src/OrbitBase/SimpleExecutor.cpp:11]: MutexLock

File: src/OrbitBase/ThreadPool.cpp
Comment (beckerhe) [reply src/OrbitBase/ThreadPool.cpp:197]: This is explained in the commit message. `Executor` - the new base class - uses `std::enable_shared_from_this`.

File: src/OrbitBase/SimpleExecutor.cpp
Comment (beckerhe) [reply src/OrbitBase/SimpleExecutor.cpp:11]: Ahhh yes. Thanks for spotting this. Fix: #2411

File: src/OrbitBase/ThreadPool.cpp
Comment (dimitry-) [reply src/OrbitBase/ThreadPool.cpp:197]: Maybe add a comment here as well, so that reader doesn't hunt for commit message. Is is far enough away from ScheduleAfter...

Comment (beckerhe) [reply src/OrbitBase/ThreadPool.cpp:197]: Done. See #2411

**Pull request merged**
